[BUG] ClassNotFoundException when debugging tests in a maven multimodule project

NeoVim Version
v0.11.4
neotest-java version: >= v0.22.2

When debugging tests in a maven multimodule project, the junit debug command errors with ClassNotFoundException.

Steps to reproduce the behavior:

git clone https://github.com/LorenzoBettini/maven-multimodule-example
nvim maven-multimodule-example
open my-app-parent/my-app-module1/src/test/java/org/my/app/module1/AppTest.java and debug test
Notification: `./target/classes/junit-reports/191025234355 is not accessible by the current user!`
I am getting following error:

```
cat /home/ml_tx/.local/state/nvim/neotest-java.log 
DEBUG | 2025-10-19T23:43:55Z+0200 | ...vim/lazy/neotest-java/lua/neotest-java/types/project.lua:27 | Searching for project files:  pom.xml
DEBUG | 2025-10-19T23:43:55Z+0200 | ...vim/lazy/neotest-java/lua/neotest-java/types/project.lua:28 | Root directory:  /tmp/maven-multimodule-example
DEBUG | 2025-10-19T23:43:55Z+0200 | ...vim/lazy/neotest-java/lua/neotest-java/types/project.lua:40 | Found project directories:  { "/tmp/maven-multimodule-example/my-app-parent/pom.xml", "/tmp/maven-multimodule-example/my-app-parent/
my-app-module1/pom.xml", "/tmp/maven-multimodule-example/my-app-parent/my-app-module2/pom.xml" }
DEBUG | 2025-10-19T23:43:55Z+0200 | ...vim/lazy/neotest-java/lua/neotest-java/types/project.lua:57 | modules:  { "/tmp/maven-multimodule-example/my-app-parent", "/tmp/maven-multimodule-example/my-app-parent/my-app-module1", "/tmp/mav
en-multimodule-example/my-app-parent/my-app-module2" }
DEBUG | 2025-10-19T23:43:55Z+0200 | ...t-java/lua/neotest-java/util/find_module_by_filepath.lua:7 | module_dirs { "/tmp/maven-multimodule-example/my-app-parent", "/tmp/maven-multimodule-example/my-app-parent/my-app-module1", "/tmp/ma
ven-multimodule-example/my-app-parent/my-app-module2" }
DEBUG | 2025-10-19T23:43:55Z+0200 | ...t-java/lua/neotest-java/util/find_module_by_filepath.lua:8 | filepath /tmp/maven-multimodule-example/my-app-parent/my-app-module1/src/test/java/org/my/app/module1/AppTest.java
DEBUG | 2025-10-19T23:43:55Z+0200 | ...-java/lua/neotest-java/command/junit_command_builder.lua:78 | assigned basedir: /tmp/maven-multimodule-example/my-app-parent/my-app-module1
DEBUG | 2025-10-19T23:43:55Z+0200 | ...va/lua/neotest-java/core/spec_builder/compiler/jdtls.lua:55 | compilation in incremental mode
DEBUG | 2025-10-19T23:43:55Z+0200 | ...va/lua/neotest-java/core/spec_builder/compiler/jdtls.lua:69 | compilation complete!
DEBUG | 2025-10-19T23:43:55Z+0200 | ...va/lua/neotest-java/core/spec_builder/compiler/jdtls.lua:71 | scanning for test resources in /tmp/maven-multimodule-example/my-app-parent/my-app-module1
DEBUG | 2025-10-19T23:43:55Z+0200 | ...vim/lazy/neotest-java/lua/neotest-java/command/jdtls.lua:102 | classpath entries: 6
DEBUG | 2025-10-19T23:43:55Z+0200 | ...m/lazy/neotest-java/lua/neotest-java/util/write_file.lua:16 | writing to file:  ./target/classes/.cp
DEBUG | 2025-10-19T23:43:55Z+0200 | ...neotest-java/lua/neotest-java/core/spec_builder/init.lua:84 | junit debug command:  /usr/lib/jvm/java-25-openjdk/bin/java   -agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=0.0.0.0:
43791 -Xdebug -Dspring.config.additional-location=optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/app
lication.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application-test.yml,optional:fi
le:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application-test.properties,optional:file:/tmp/maven-multi
module-example/my-app-parent/target/classes/test-classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/test-classes/application.yaml,optional:file:/tmp/maven-multimodule-example/my-app-pare
nt/target/classes/test-classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/test-classes/application-test.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/class
es/test-classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/test-classes/application-test.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target
/classes/classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/classes/application.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/class
es/classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/classes/application-test.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/
classes/classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/classes/application-test.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-modu
le1/target/classes/test-classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/test-classes/application.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-
module1/target/classes/test-classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/test-classes/application-test.yml,optional:file:/tmp/maven-multimodule-example/my-app
-parent/my-app-module1/target/classes/test-classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/test-classes/application-test.properties,optional:file:/tmp/maven-multi
module-example/my-app-parent/my-app-module2/target/classes/classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application.yaml,optional:file:/tmp/maven-multimodule
-example/my-app-parent/my-app-module2/target/classes/classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application-test.yml,optional:file:/tmp/maven-multim
odule-example/my-app-parent/my-app-module2/target/classes/classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application-test.properties,optional:file:/tmp/m
aven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application.yaml,optional:file:/t
mp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application-test.yml,o
ptional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/applicat
ion-test.properties -jar /home/ml_tx/.local/share/nvim/neotest-java/junit-platform-console-standalone-1.10.1.jar execute @./target/classes/.cp --reports-dir=./target/classes/junit-reports/191025234355 --fail-if-no-tests --disable-ban
ner --details=testfeed --config=junit.platform.output.capture.stdout=true --select-method='org.my.app.module1.AppTest#testApp'
TRACE | 2025-10-19T23:43:55Z+0200 | ...m/lazy/neotest-java/lua/neotest-java/build_tool/init.lua:31 | run_debug_test function
DEBUG | 2025-10-19T23:43:55Z+0200 | ...m/lazy/neotest-java/lua/neotest-java/build_tool/init.lua:58 | starting job with command:  /usr/lib/jvm/java-25-openjdk/bin/java   -agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=0.
0.0.0:43791 -Xdebug -Dspring.config.additional-location=optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/class
es/application.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application-test.yml,optio
nal:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/classes/application-test.properties,optional:file:/tmp/maven
-multimodule-example/my-app-parent/target/classes/test-classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/test-classes/application.yaml,optional:file:/tmp/maven-multimodule-example/my-ap
p-parent/target/classes/test-classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/test-classes/application-test.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target
/classes/test-classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/target/classes/test-classes/application-test.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/
target/classes/classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/classes/application.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target
/classes/classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/classes/application-test.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/t
arget/classes/classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/classes/application-test.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-ap
p-module1/target/classes/test-classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/test-classes/application.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/m
y-app-module1/target/classes/test-classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/test-classes/application-test.yml,optional:file:/tmp/maven-multimodule-example/
my-app-parent/my-app-module1/target/classes/test-classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module1/target/classes/test-classes/application-test.properties,optional:file:/tmp/maven
-multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application.yaml,optional:file:/tmp/maven-multi
module-example/my-app-parent/my-app-module2/target/classes/classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application-test.yml,optional:file:/tmp/maven-
multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/classes/application-test.properties,optional:file:
/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application.yaml,optional:f
ile:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application.properties,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application-test
.yml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/application-test.yaml,optional:file:/tmp/maven-multimodule-example/my-app-parent/my-app-module2/target/classes/test-classes/ap
plication-test.properties -jar /home/ml_tx/.local/share/nvim/neotest-java/junit-platform-console-standalone-1.10.1.jar execute @./target/classes/.cp --reports-dir=./target/classes/junit-reports/191025234355 --fail-if-no-tests --disab
le-banner --details=testfeed --config=junit.platform.output.capture.stdout=true --select-method='org.my.app.module1.AppTest#testApp'
DEBUG | 2025-10-19T23:43:56Z+0200 | ...zy/neotest-java/lua/neotest-java/core/result_builder.lua:62 | Found report files:  {}
DEBUG | 2025-10-19T23:43:56Z+0200 | ...m/lazy/neotest-java/lua/neotest-java/build_tool/init.lua:51 | command exited with code:  255
ERROR | 2025-10-19T23:43:56Z+0200 | ...m/lazy/neotest-java/lua/neotest-java/build_tool/init.lua:53 | command exited with code:  255
ERROR | 2025-10-19T23:43:56Z+0200 | ...m/lazy/neotest-java/lua/neotest-java/build_tool/init.lua:54 | stderr:  OpenJDK 64-Bit Server VM warning: Option -Xdebug was deprecated in JDK 22 and will likely be removed in a future release.
org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-jupiter' failed to discover tests
        at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:160)
        at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverSafely(EngineDiscoveryOrchestrator.java:132)
        at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:107)
        at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:78)
        at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:99)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:85)
        at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:47)
        at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:63)
        at org.junit.platform.console.tasks.ConsoleTestExecutor.executeTests(ConsoleTestExecutor.java:102)
        at org.junit.platform.console.tasks.ConsoleTestExecutor.lambda$execute$1(ConsoleTestExecutor.java:71)
        at org.junit.platform.console.tasks.CustomContextClassLoaderExecutor.replaceThreadContextClassLoaderAndInvoke(CustomContextClassLoaderExecutor.java:42)
        at org.junit.platform.console.tasks.CustomContextClassLoaderExecutor.invoke(CustomContextClassLoaderExecutor.java:33)
        at org.junit.platform.console.tasks.ConsoleTestExecutor.execute(ConsoleTestExecutor.java:71)
        at org.junit.platform.console.options.ExecuteTestsCommand.execute(ExecuteTestsCommand.java:61)
        at org.junit.platform.console.options.ExecuteTestsCommand.execute(ExecuteTestsCommand.java:27)
        at org.junit.platform.console.options.BaseCommand.call(BaseCommand.java:74)
        at org.junit.platform.console.shadow.picocli.CommandLine.executeUserObject(CommandLine.java:2041)
        at org.junit.platform.console.shadow.picocli.CommandLine.access$1500(CommandLine.java:148)
        at org.junit.platform.console.shadow.picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
        at org.junit.platform.console.shadow.picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
        at org.junit.platform.console.shadow.picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
        at org.junit.platform.console.shadow.picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
        at org.junit.platform.console.shadow.picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
        at org.junit.platform.console.shadow.picocli.CommandLine.execute(CommandLine.java:2170)
        at org.junit.platform.console.options.MainCommand.runCommand(MainCommand.java:137)
        at org.junit.platform.console.options.MainCommand.run(MainCommand.java:125)
        at org.junit.platform.console.options.CommandFacade.run(CommandFacade.java:36)
        at org.junit.platform.console.ConsoleLauncher.run(ConsoleLauncher.java:57)
        at org.junit.platform.console.ConsoleLauncher.run(ConsoleLauncher.java:42)
        at org.junit.platform.console.ConsoleLauncher.main(ConsoleLauncher.java:35)
Caused by: org.junit.platform.commons.JUnitException: MethodSelector [className = ''org.my.app.module1.AppTest', methodName = 'testApp'', parameterTypes = '', classLoader = null] resolution failed
        at org.junit.platform.launcher.listeners.discovery.AbortOnFailureLauncherDiscoveryListener.selectorProcessed(AbortOnFailureLauncherDiscoveryListener.java:39)
        at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:103)
        at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.run(EngineDiscoveryRequestResolution.java:83)
        at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolver.resolve(EngineDiscoveryRequestResolver.java:113)
        at org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:46)
        at org.junit.jupiter.engine.JupiterTestEngine.discover(JupiterTestEngine.java:69)
        at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:152)
        ... 29 more
Caused by: org.junit.platform.commons.PreconditionViolationException: Could not load class with name: 'org.my.app.module1.AppTest
        at org.junit.platform.engine.discovery.MethodSelector.lambda$lazyLoadJavaClass$0(MethodSelector.java:239)
        at org.junit.platform.commons.function.Try$Failure.getOrThrow(Try.java:335)
        at org.junit.platform.engine.discovery.MethodSelector.lazyLoadJavaClass(MethodSelector.java:238)
        at org.junit.platform.engine.discovery.MethodSelector.getJavaClass(MethodSelector.java:193)
        at org.junit.jupiter.engine.discovery.MethodSelectorResolver.resolve(MethodSelectorResolver.java:75)
        at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.lambda$resolve$2(EngineDiscoveryRequestResolution.java:150)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1693)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:147)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:588)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:574)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
        at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
        at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:683)
        at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:189)
        at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:126)
        at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDi
        ... 34 more
Caused by: java.lang.ClassNotFoundException: 'org.my.app.module1.AppTest
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:377)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:557)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:547)
        at org.junit.platform.commons.util.ReflectionUtils.lambda$tryToLoadClass$9(ReflectionUtils.java:832)
        at org.junit.platform.commons.function.Try.lambda$call$0(Try.java:57)
        at org.junit.platform.commons.function.Try.of(Try.java:93)
        at org.junit.platform.commons.function.Try.call(Try.java:57)
        at org.junit.platform.commons.util.ReflectionUtils.tryToLoadClass(ReflectionUtils.java:795)
        at org.junit.platform.commons.util.ReflectionUtils.tryToLoadClass(ReflectionUtils.java:751)
        at org.junit.platform.engine.discovery.MethodSelector.lazyLoadJavaClass(MethodSelector.java:236)
        ... 49 more
```

--- a/lua/neotest-java/command/jdtls.lua
+++ b/lua/neotest-java/command/jdtls.lua
@@ -105,11 +105,9 @@ M.get_classpath = function(additional_classpath_entries, dir)
 end
 
 M.get_classpath_file_argument = function(report_dir, additional_classpath_entries, dir)
-	local classpath = table.concat(M.get_classpath(additional_classpath_entries, dir), ":")
-	local temp_file = compatible_path(report_dir .. "/.cp")
-	write_file(temp_file, ("-cp %s"):format(classpath))
-
-	return ("@%s"):format(temp_file)
+	local classpaths = M.get_classpath(additional_classpath_entries, dir)
+	local classpath = table.concat(classpaths, ":")
+	return classpath
 end
 
 return M

--- a/lua/neotest-java/command/junit_command_builder.lua
+++ b/lua/neotest-java/command/junit_command_builder.lua
@@ -99,11 +99,15 @@ local CommandBuilder = {
 		local selectors = {}
 		for _, v in ipairs(self._test_references) do
 			if v.type == "test" then
-				table.insert(selectors, "--select-method='" .. v.qualified_name .. "'")
+				local class_name = v.qualified_name:match("^(.-)#") or v.qualified_name
+				table.insert(selectors, "--select-class=" .. class_name)
+				if v.method_name then
+					table.insert(selectors, "--select-method=" .. v.method_name)
+				end
 			elseif v.type == "file" then
 				table.insert(selectors, "-c=" .. v.qualified_name)
 			elseif v.type == "dir" then
-				selectors = "-p=" .. v.qualified_name
+				table.insert(selectors, "-p=" .. v.qualified_name)
 			end
 		end
 		assert(#selectors ~= 0, "junit command has to have a selector")
@@ -115,7 +119,7 @@ local CommandBuilder = {
 				"-jar",
 				self._junit_jar,
 				"execute",
-				("%s"):format(self._classpath_file_arg),
+				"--classpath=" .. self._classpath_file_arg,
 				"--reports-dir=" .. self._reports_dir,
 				"--fail-if-no-tests",
 				"--disable-banner",

